src/BSON/Iterator.c

```diff
@@ -0,0 +1,252 @@
+/*
+ * Copyright 2014-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <php.h>
+#include <ext/standard/base64.h>
+#include <Zend/zend_interfaces.h>
+#include <Zend/zend_operators.h>
+#include <ext/standard/php_var.h>
+#include <zend_smart_str.h>
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "php_phongo.h"
+#include "phongo_error.h"
+#include "BSON/Iterator.h"
+#include "BSON/Iterator_arginfo.h"
+
+zend_class_entry* php_phongo_iterator_ce;
+
+static bool php_phongo_iterator_init_with_zval(php_phongo_iterator_t* iterator, zval* zbson)
+{
+	php_phongo_document_t*  document_intern;
+	php_phongo_arraylist_t* arraylist_intern;
+	const bson_t*           bson;
+
+	ZVAL_COPY(&iterator->bson, zbson);
+	if (instanceof_function(Z_OBJCE_P(zbson), php_phongo_document_ce)) {
+		document_intern = Z_DOCUMENT_OBJ_P(&iterator->bson);
+		bson            = document_intern->bson;
+	} else if (instanceof_function(Z_OBJCE_P(zbson), php_phongo_arraylist_ce)) {
+		arraylist_intern   = Z_ARRAYLIST_OBJ_P(&iterator->bson);
+		bson               = arraylist_intern->bson;
+		iterator->is_array = true;
+	} else {
+		phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT, "Could not create iterator for %s instance.", PHONGO_ZVAL_CLASS_OR_TYPE_NAME_P(zbson));
+
+		return false;
+	}
+
+	if (!bson_iter_init(&iterator->iter, bson)) {
+		phongo_throw_exception(PHONGO_ERROR_UNEXPECTED_VALUE, "Could not create iterator for BSON instance.");
+
+		return false;
+	}
+
+	iterator->valid = bson_iter_next(&iterator->iter);
```

Comment:
As you noted it's consistent with other iterators, but I'll caution that this works because we allow users to call `rewind` as long as we haven't advanced beyond the first key.

Reply:
I don't follow how this is related to original restrictions on `rewind()`, but is there any conflict with changes to allow `rewind()` to be called at any point (as I suggested in https://github.com/mongodb/mongo-php-driver/pull/1387#discussion_r1040295247)?

Reply:
> If we were to forbid calling rewind() entirely (even when we haven't advanced yet), this would entirely break the iteration logic.

I don't think this was ever relevant to PHPC or PHPLIB. Cursor (and ChangeStream by composition) is the closest thing, and that only disallows `rewind()` after advancing beyond the first element.

Reply:
Feel free to resolve after reading.